compiler/src/main/kotlin/io/spine/protodata/codegen/BaseTypeSystem.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2023, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata.codegen
+
+import io.spine.protodata.PrimitiveType
+import io.spine.protodata.Type
+import io.spine.protodata.TypeName
+import io.spine.protodata.Value
+import io.spine.protodata.Value.KindCase.BOOL_VALUE
+import io.spine.protodata.Value.KindCase.BYTES_VALUE
+import io.spine.protodata.Value.KindCase.DOUBLE_VALUE
+import io.spine.protodata.Value.KindCase.ENUM_VALUE
+import io.spine.protodata.Value.KindCase.INT_VALUE
+import io.spine.protodata.Value.KindCase.LIST_VALUE
+import io.spine.protodata.Value.KindCase.MAP_VALUE
+import io.spine.protodata.Value.KindCase.MESSAGE_VALUE
+import io.spine.protodata.Value.KindCase.NULL_VALUE
+import io.spine.protodata.Value.KindCase.STRING_VALUE
+
+/**
+ * The type system of an application being built by ProtoData.
+ *
+ * The type system knows the association between the Protobuf types and the types of
+ * a target language.
+ *
+ * @param T the type of a language-specific type name
+ * @param V the type of a language-specific expression, i.e. the code that yields a value
+ */
+public abstract class BaseTypeSystem<T: CodePrintable, V: CodePrintable>
+protected constructor(
+    private val knownTypes: Map<TypeName, T>
+) {
+
+    /**
+     * A [ValueConverter] for converting Protobuf values into language-specific code expressions.
+     */
+    protected abstract val valueConverter: ValueConverter<V>
+
+    /**
+     * Converts the given Protobuf type into a language-specific type name.
+     */
+    public fun typeNameToCode(type: Type): T {
```

Comment:
I don't think this method name fits the definition of the generic parameter `T`. `T` is a type name. We probably can obtain an expression out of `T`, so that it becomes `V` (while both are `CodePrintable`, which introduce additional confusion). But it's another story.

Reply:
If we convert a Proto type (name) to a language-specific type, I don't understand why we need a type system. The conversion is governed by code generation process for this kind of proto type and target programming language.

We don't need to gather all the types known to the application for this conversion. It's a matter of combination of two things: source Proto type and target programming language.

Reply:
I'm not against gathering all known types. We need this. But we don't do this for the conversion.

Reply:
We do need a type system to convert Proto names into Java named. That's the whole reason `TypeSystem` hierarchy exists.

Since we cannot always drag the whole info about the `.proto` file with us, and since we need that info to know how the generated type (e.g. Java class) will turn out, we need some kind of way to keep track of all the types.

For example, the `Type` in `fun typeNameToCode(type: Type)` knows nothing about outer classes and Java packages, yet it is a unique identifier for a Proto type.